Assets/HoloToolkit/Interaction/Scripts/InteractibleObject.cs

```diff
@@ -0,0 +1,46 @@
+﻿//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+//
+using System;
+using UnityEngine;
+using HoloToolkit.Unity.InputModule;
+
+namespace HoloToolkit.Unity.Interaction
+{
+    /// <summary>
+    /// Objects wanting to recieve messages can inherit from this class.
+    /// The filtertag still needs to be implemented on the FocusManager/InputManager side to 
+    /// Match functionality from MRDL.
+    /// </summary>
+    public abstract class InteractibleObject : MonoBehaviour, IPointerSpecificFocusable, IInputHandler, IHoldHandler, IManipulationHandler, INavigationHandler, ISourceStateHandler
```

Comment:
Do we really need this class? Shouldn't the interactibles just implement what they need? We don't need clear references to interfaces in the InputManager since it uses Unitys event system that sends events to gameObjects which will then find the interface references on its own.

Reply:
The basic problem has been with the discoverability of interfaces for what interactions an object can consume.  This would be better solved on the InputManager front by knowing what event's I can consume as oppose to the guess work of going through all the interfaces especially for global listeners.  

The InteractibleObject just says I'm and easy base class that takes all interactions and you can easily find what you want to override by simply extending and override the function.  

Currently there's not an easy way to filter object interactions and or provide some base functionality for things you interact with that's common across all of them.  I'm open to other alternatives to provide filtering methods and other logic commonly across all interaction objects.  Perhaps this should be refactored in InputManager instead and provide better delegate hooks as opposed to just global listeners?

Thoughts?

thanks,
-pat

Reply:
>The basic problem has been with the discoverability of interfaces for what interactions an object can consume. This would be better solved on the InputManager front by knowing what event's I can consume as oppose to the guess work of going through all the interfaces especially for global listeners.

@paseb I'm not sure if I understand this.  What do you mean by "the discoverability of interfaces"?  From what I understand the reason why we're implementing the Interfaces was to get away from delegates and events to improve overall system performance.

Reply:
@StephenHodgson, That performance is negligible. We have a slider with one side being general usability and discoverability and the other considered optimal performance.  I can handwrite a shader or write an app completely in assembly to get the best performance, though it would be crap for anyone to try to take that over or work off of what I've done.

Having just interfaces without the ability to quickly use intellisense to see what I can hook into from events represents the same problem.  That's why we need some middle ground and if someone wants to implement interfaces in all their base classes that fine.  Also if someone wanted to just extend an InteractibleObject and override an virtual function that should be fine as well.

I prefer delegates or event/actions with names variables and really haven't seen any perf issues.  Has anyone done a hard study on the actual per impacts?

thanks,
-pat

Reply:
The Unity event system does come with a overhead because it really just does a `GetComponent<Interface>` in the background.
Here are some real basic profiler results run in the editor with ~30 and ~100 global listeners.
![hwphjox](https://user-images.githubusercontent.com/5775094/30833077-cea35ec4-a24d-11e7-8ae4-8fd80270a408.png)
![smcdkdn](https://user-images.githubusercontent.com/5775094/30833086-d8665fce-a24d-11e7-8578-9cd66b2d47fa.png)

Reply:
@Zod- Not sure what you have hooked up behind the 2 images above.  To be clear I wasn't talking about the Unity "sendEvent" event system but the c# tightly typed event delegates as such:
'''public static event Action<GameObject, InteractionEventArgs> OnTapped;'''
For me the discoverability (intellisense and autocomplete of the callback) of what actions 
I can subscribe to outweighs any perf considerations.  Unless you're registering and unregistering a massive number of delegates you shouldn't notice any performance hit.

thanks,
-pat

Reply:
> That performance is negligible.

I'd like to disagree, it was a major part of the reason the input system was refactored.  Although, I do have to agree with your analogy. The difference being that there's exposed interfaces for people to tap into and use as they need.

I think the biggest issue here is that most people don't fully understand it, and how to use it properly.

> Having just interfaces without the ability to quickly use intellisense to see what I can hook into from events represents the same problem.

I agree, this is somewhat of an issue, and I can see the benefit to how this was arranged to make finding interactions easier.  This goes back to the previous point.

> That's why we need some middle ground and if someone wants to implement interfaces in all their base classes that fine. Also if someone wanted to just extend an `InteractibleObject` and override an virtual function that should be fine as well.

I think I better understand what this class is supposed to do now, thanks.  It's more of a basic building block for overriding and implementing custom logic.

>I prefer delegates or event/actions with names variables and really haven't seen any perf issues. 
 Has anyone done a hard study on the actual perf impacts?

Yes, there was a whole thread covering the tradeoffs and perf impacts with each type we suggested.  The biggest push was getting away from `GameObject.SendMessage`.

Subscribing to delegate events decreases perf logarithmically, see dramatic impact as more objects subscribe.

Reply:
@paseb as far as perf is concerned, here's a link to a good perf test you can do on your own:
http://jacksondunstan.com/articles/3043

Reply:
@StephenHodgson, Thanks for the quick response!  I'm totally fine finding a middle ground were we can provide simple discoverability and fundamental building blocks that are easy to extend.  That was the whole purpose behind MRDL and HUX.  Getting away from `GameObject.SendMessage` was the right thing to do.  We provide both methods in MRDL to allow creators to leverage whatever method was easier for them though delegates were the easiest for quickly finding all the events you could plug into on a component.

Are there other ways we could have a set of building block abstract classes that provide a foundation to rapidly develop off of?

thanks,
-pat

Reply:
> Are there other ways we could have a set of building block abstract classes that provide a foundation to rapidly develop off of?

Not that I can think of, you're on the right track.

Reply:
I was just looking at Jackson's other articles as well.  Crazy sauce.

So in summary:
I'm not sure we actually need InteractibleObject as a base abstract class but I'm not sure there is any other alternative that would provide the same easy override functionality.  A lot of the MRDL examples and base functionality uses InteractibleObject as a foundation class to build from.  I'm open to refactoring it moving forward though think we need it to integrate other MRDL functionality.  

thanks,
-pat

Reply:
Whatever event system we will have in the end, an important thing to keep in mind is that inheriting from `InteractibleObject` has other side-effects than just implement a single interface. 

I've recently run into the issue that adding a gameObject as modal will only absorb the interfaces that it implements and I needed it to lock them all. `InteractibleObject` will do this out of the box as it implements all interfaces but some may find this unexpected behavior?

Reply:
Agreed, let's not forget that Unity plans on doing a whole revamp of the Input System as well in a near future release.  That's something we also need to take into consideration.

Reply:
@StephenHodgson,
"At some point devs will need to make the custom logic themselves for their implementations in their own apps. Giving the option and examples of what to use is better than packing it all in without a choice."

I'm not worried about devs as they should be fine using and finding the interfaces.  They are not the only ones using the toolkit and currently the toolkit is really only targeting devs.  I'm more focused on designers, tech artists, creative coders and anyone else that want get off the ground quickly and have an extensible set of abstract classes to build off of.  That's sort of the purpose of MRDL and bringing some functionality over to MRTK.

@Zod- InteractibleObject isn't meant to be the one ring but simply a starting point to extend from.  Any dev would most likely start with base classes that just implement the interface they want anyway and not start with InteractibleObject.  I'm completely open to other implementations if they meet the same goal for accessibility.

Again awesome input on all fronts!

thanks,
-pat

Reply:
Given that the modal issue and that the input system will be changing on the Unity side in the near future I'll remove InteractibleObject in favor of specifically targeted interactibles starting just with the base button class.  I think we also need a better way to subscribe to interactions through input manager but can look at that later.

thanks,
-pat